File: src/LTAdmin/Services/Business/ExamService.cs
using System.Data;
using LTAdmin.Core;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;
using LTAdmin.Repositories;
using LTAdmin.Services.Common;
using LTAdmin.Services.Logging;
using LTAdmin.Services.Referentiel;
using LTAdmin.Services.Validation;

namespace LTAdmin.Services.Business;

/// <summary>
/// Examens : sessions, épreuves, notes d'examen et résultats finaux.
/// Réimplémentation C# de GENERER_RESULTAT_FINAL(idClasse, idSession) :
/// moyenne = CC × POIDS_CC/100 + examen × POIDS_EXAMEN/100, mention via
/// GRILLE_MENTION, décision via MOY_ADMISSION, rang. Idempotent par
/// inscription (le schéma ne porte pas de colonne session sur RESULTAT_FINAL).
/// </summary>
public sealed class ExamService : ServiceBase
{
    private readonly ExamRepository _examens;
    private readonly BulletinRepository _bulletins;
    private readonly EnrollmentRepository _inscriptions;
    private readonly ReferentielRepository _referentiel;

    public ExamService(AccessDatabase database, AppLogger logger, JournalService journal, ParametreService parametres)
        : base(database, logger, journal, parametres)
    {
        _examens = new ExamRepository(database);
        _bulletins = new BulletinRepository(database);
        _inscriptions = new EnrollmentRepository(database);
        _referentiel = new ReferentielRepository(database);
    }

    public List<SessionExam> ListSessions(int? idAnnee = null)
    {
        try { return _examens.ListSessions(idAnnee); }
        catch (Exception ex)
        {
            Logger.Error("Liste des sessions d’examen", ex);
            return new List<SessionExam>();
        }
    }

    public Result<int> CreateSession(SessionExam session, string codeUtr)
    {
        var validation = ExamValidator.ValidateSession(session);
        if (!validation.IsValid) return Invalid<int>(validation);
        try
        {
            var id = _examens.InsertSession(session);
            Journal.LogCreation(codeUtr, Tables.SessionExam, id, session.Libelle);
            return Result<int>.Ok(id, "Session d’examen créée.");
        }
        catch (Exception ex) { return Failure<int>("Création d’une session", ex); }
    }

    public Result SetSessionCloturee(int idSession, bool cloturee, string codeUtr)
    {
        try
        {
            var session = _examens.GetSession(idSession);
            if (session is null)
                return Result.Fail("Session introuvable.", ErrorCodes.NotFound);
            _examens.SetSessionCloturee(idSession, cloturee);
            Journal.LogOperation(codeUtr, cloturee ? "CLOTURE_SESSION" : "REOUVERTURE_SESSION",
                Tables.SessionExam, idSession, session.Libelle);
            return Result.Ok(cloturee ? "Session clôturée." : "Session rouverte.");
        }
        catch (Exception ex) { return Failure("Clôture d’une session", ex); }
    }

    public List<EpreuveDetail> ListEpreuves(int? idSession = null, int? idClasse = null)
    {
        try { return _examens.ListEpreuves(idSession, idClasse); }
        catch (Exception ex)
        {
            Logger.Error("Liste des épreuves", ex);
            return new List<EpreuveDetail>();
        }
    }

    public Result<int> CreateEpreuve(Epreuve epreuve, string codeUtr)
    {
        if (!epreuve.Bareme.HasValue || epreuve.Bareme.Value <= 0)
            epreuve.Bareme = Parametres.BaremeDefaut;
        if (!epreuve.Coefficient.HasValue || epreuve.Coefficient.Value <= 0)
            epreuve.Coefficient = 1d;
        var validation = ExamValidator.ValidateEpreuve(epreuve);
        if (!validation.IsValid) return Invalid<int>(validation);
        try
        {
            var session = _examens.GetSession(epreuve.IdSession!.Value);
            if (session is null)
                return Result<int>.Fail("Session introuvable.", ErrorCodes.NotFound);
            if (session.Cloturee == true)
                return Result<int>.Fail("Session clôturée : rouvrez-la pour ajouter une épreuve.", ErrorCodes.BusinessRule);
            var id = _examens.InsertEpreuve(epreuve);
            Journal.LogCreation(codeUtr, Tables.Epreuve, id, epreuve.CodeMatiere);
            return Result<int>.Ok(id, "Épreuve créée.");
        }
        catch (Exception ex) { return Failure<int>("Création d’une épreuve", ex); }
    }

    public Result UpdateEpreuve(Epreuve epreuve, string codeUtr)
    {
        if (!epreuve.IdEpreuve.HasValue)
            return Result.Fail("Épreuve introuvable.", ErrorCodes.NotFound);
        var validation = ExamValidator.ValidateEpreuve(epreuve);
        if (!validation.IsValid) return Invalid(validation);
        try
        {
            var updated = _examens.UpdateEpreuve(epreuve);
            if (updated == 0)
                return Result.Fail("Épreuve introuvable.", ErrorCodes.NotFound);
            Journal.LogModification(codeUtr, Tables.Epreuve, epreuve.IdEpreuve, epreuve.CodeMatiere);
            return Result.Ok("Épreuve enregistrée.");
        }
        catch (Exception ex) { return Failure("Modification d’une épreuve", ex); }
    }

    public Result DeleteEpreuve(int idEpreuve, string codeUtr)
    {
        try
        {
            var existing = _examens.GetEpreuve(idEpreuve);
            if (existing is null)
                return Result.Fail("Épreuve introuvable.", ErrorCodes.NotFound);
            _examens.DeleteEpreuve(idEpreuve);
            Journal.LogSuppression(codeUtr, Tables.Epreuve, idEpreuve, existing.CodeMatiere);
            return Result.Ok("Épreuve supprimée.");
        }
        catch (Exception ex)
        {
            if (OleDbExceptionHelper.IsForeignKeyViolation(ex))
                return Result.Fail("Suppression impossible : des notes d’examen sont déjà saisies.", ErrorCodes.ForeignKeyViolation);
            return Failure("Suppression d’une épreuve", ex);
        }
    }

    public List<NoteSaisieRow> GetGrilleSaisie(int idEpreuve)
    {
        try
        {
            var epreuve = _examens.GetEpreuveDetail(idEpreuve);
            if (epreuve?.IdClasse is null) return new List<NoteSaisieRow>();
            return _examens.ListGrilleSaisie(idEpreuve, epreuve.IdClasse.Value);
        }
        catch (Exception ex)
        {
            Logger.Error("Grille de saisie des notes d’examen", ex);
            return new List<NoteSaisieRow>();
        }
    }

    public Result UpsertNoteExamen(int idEpreuve, int idInscription, double? valeur, bool absent, string? copieNum, string codeUtr)
    {
        try
        {
            var epreuve = _examens.GetEpreuveDetail(idEpreuve);
            if (epreuve is null)
                return Result.Fail("Épreuve introuvable.", ErrorCodes.NotFound);
            var bareme = epreuve.Bareme ?? Parametres.BaremeDefaut;
            var validation = GradeValidator.ValidateNote(valeur, absent, bareme);
            if (!validation.IsValid) return Invalid(validation);
            if (!string.IsNullOrEmpty(copieNum) && copieNum.Length > 40)
                return Result.Fail("N° de copie : 40 caractères maximum.", ErrorCodes.Validation);

            var inscription = _inscriptions.GetById(idInscription);
            if (inscription is null)
                return Result.Fail("Inscription introuvable.", ErrorCodes.NotFound);
            if (inscription.IdClasse != epreuve.IdClasse)
                return Result.Fail("Cet étudiant n’est pas inscrit dans la classe de cette épreuve.", ErrorCodes.BusinessRule);

            var existing = _examens.GetNoteExamen(idEpreuve, idInscription);
            if (existing is null)
            {
                var id = _examens.InsertNoteExamen(new NoteExamen
                {
                    IdEpreuve = idEpreuve,
                    IdInscription = idInscription,
                    ValeurNote = absent ? null : valeur,
                    Absent = absent,
                    CopieNum = copieNum,
                    DateSaisie = DateTime.Now,
                    CodeUtr = codeUtr
                });
                Journal.LogCreation(codeUtr, Tables.NoteExamen, id, $"Épreuve {idEpreuve} / inscrit {idInscription}");
            }
            else
            {
                existing.ValeurNote = absent ? null : valeur;
                existing.Absent = absent;
                existing.CopieNum = copieNum;
                existing.DateSaisie = DateTime.Now;
                existing.CodeUtr = codeUtr;
                _examens.UpdateNoteExamen(existing);
                Journal.LogModification(codeUtr, Tables.NoteExamen, existing.IdNoteEx, $"Épreuve {idEpreuve} / inscrit {idInscription}");
            }
            return Result.Ok("Note d’examen enregistrée.");
        }
        catch (Exception ex) { return Failure("Saisie d’une note d’examen", ex); }
    }

    public List<ResultatFinal> ListResultats(int idClasse)
    {
        try { return _bulletins.ListByClasse(idClasse); }
        catch (Exception ex)
        {
            Logger.Error("Liste des résultats finaux", ex);
            return new List<ResultatFinal>();
        }
    }

    /// <summary>
    /// Génère les résultats finaux d'une classe pour une session.
    /// MOY_CC = moyenne des bulletins (toutes périodes) ; MOY_EXAM = moyenne
    /// pondérée des notes d'examen normalisées sur 20 ; si l'une des deux
    /// parties est absente, la moyenne générale reprend la partie disponible.
    /// </summary>
    public Result<int> GenererResultatsFinaux(int idClasse, int idSession, string codeUtr, string? observation = null)
    {
        try
        {
            var classe = _referentiel.GetClasse(idClasse);
            if (classe is null)
                return Result<int>.Fail("Classe introuvable.", ErrorCodes.NotFound);
            var session = _examens.GetSession(idSession);
            if (session is null)
                return Result<int>.Fail("Session introuvable.", ErrorCodes.NotFound);
            if (session.Cloturee == true)
                return Result<int>.Fail("Session clôturée : rouvrez-la pour régénérer les résultats.", ErrorCodes.BusinessRule);

            var inscrits = _inscriptions.ListByClasse(idClasse);
            if (inscrits.Count == 0)
                return Result<int>.Fail($"Aucun inscrit dans la classe « {classe.Libelle} ».", ErrorCodes.BusinessRule);

            var grille = _bulletins.ListMentions();
            var poidsCc = Parametres.PoidsCc;
            var poidsExam = Parametres.PoidsExamen;
            var seuil = Parametres.MoyAdmission;
            var moyExamParInscrit = CalculerMoyennesExamen(idClasse, idSession);

            using var tx = Db.BeginTransaction();
            var rangsSource = new List<(int Id, double? Valeur)>();
            foreach (var inscrit in inscrits)
            {
                var idInscription = inscrit.IdInscription!.Value;
                _bulletins.DeleteByInscription(idInscription);

                var moyennesBulletins = _bulletins.ListMoyennesByInscription(idInscription);
                double? moyCc = moyennesBulletins.Count > 0 ? moyennesBulletins.Average() : null;
                moyExamParInscrit.TryGetValue(idInscription, out var moyExam);

                double? moyenneGen = (moyCc, moyExam) switch
                {
                    (not null, not null) => moyCc.Value * poidsCc / 100d + moyExam.Value * poidsExam / 100d,
                    (not null, null) => moyCc,
                    (null, not null) => moyExam,
                    _ => null
                };
                var mention = MentionHelper.Trouver(grille, moyenneGen);
                var resultat = new ResultatFinal
                {
                    IdInscription = idInscription,
                    MoyCc = moyCc,
                    MoyExam = moyExam,
                    MoyenneGen = moyenneGen,
                    Mention = mention?.Mention,
                    Decision = moyenneGen.HasValue
                        ? (moyenneGen.Value >= seuil ? "ADMIS" : "AJOURNÉ")
                        : "EN ATTENTE",
                    DateDelib = DateTime.Now,
                    Observation = observation
                };
                var idResultat = _bulletins.InsertResultat(resultat);
                rangsSource.Add((idResultat, moyenneGen));
            }

            var rangs = MentionHelper.RangsCompetition(rangsSource);
            foreach (var item in rangsSource)
            {
                _bulletins.UpdateResultatRang(new ResultatFinal
                {
                    IdResultat = item.Id,
                    Rang = rangs.TryGetValue(item.Id, out var rang) && rang > 0 ? rang : null
                });
            }

            tx.Complete();
            Journal.LogOperation(codeUtr, "GENERATION_RESULTATS", Tables.ResultatFinal, null,
                $"Classe « {classe.Libelle} » × session « {session.Libelle} » : {inscrits.Count} résultat(s).");
            return Result<int>.Ok(inscrits.Count, $"{inscrits.Count} résultat(s) généré(s) pour « {classe.Libelle} ».");
        }
        catch (Exception ex) { return Failure<int>("Génération des résultats finaux", ex); }
    }

    private Dictionary<int, double?> CalculerMoyennesExamen(int idClasse, int idSession)
    {
        var table = _examens.ListNotesPonderees(idClasse, idSession);
        var cumuls = new Dictionary<int, (double Points, double Coefs)>();
        foreach (DataRow row in table.Rows)
        {
            var idInscription = DataRowMapper.GetInt32(row, "ID_INSCRIPTION");
            var valeur = DataRowMapper.GetDouble(row, "VALEUR_NOTE");
            var absent = DataRowMapper.GetBoolean(row, "ABSENT", false);
            var bareme = DataRowMapper.GetDouble(row, "BAREME");
            var coef = DataRowMapper.GetDouble(row, "COEFFICIENT") ?? 1d;
            if (!idInscription.HasValue || absent || !valeur.HasValue) continue;
            if (!bareme.HasValue || bareme.Value <= 0) continue;
            var normalisee = valeur.Value / bareme.Value * 20d;
            if (cumuls.TryGetValue(idInscription.Value, out var cumul))
                cumuls[idInscription.Value] = (cumul.Points + normalisee * coef, cumul.Coefs + coef);
            else
                cumuls[idInscription.Value] = (normalisee * coef, coef);
        }
        return cumuls.ToDictionary(
            entree => entree.Key,
            entree => entree.Value.Coefs > 0 ? (double?)(entree.Value.Points / entree.Value.Coefs) : null);
    }
}

File: ltadmin/services/business/exam_service.py
"""Gestion des examens : sessions, épreuves, saisie des notes d'examen et
calcul des résultats finaux (moyenne CC + moyenne examen pondérées)."""

from __future__ import annotations

import datetime as _dt
from typing import List, Optional

from ltadmin.core.result import Result, ResultValue
from ltadmin.data import error_helper
from ltadmin.data.access_database import AccessDatabase
from ltadmin.data.schema import Tables
from ltadmin.models.dto import EpreuveDetail, NoteSaisieRow
from ltadmin.models.entities import Epreuve, NoteExamen, ResultatFinal, SessionExam
from ltadmin.repositories.bulletin_repository import BulletinRepository
from ltadmin.repositories.enrollment_repository import EnrollmentRepository
from ltadmin.repositories.exam_repository import ExamRepository
from ltadmin.services.business.mention_helper import rangs_competition, trouver
from ltadmin.services.common import ServiceBase
from ltadmin.services.logging.app_logger import AppLogger
from ltadmin.services.logging.journal_service import JournalService
from ltadmin.services.referentiel.parametre_service import ParametreService
from ltadmin.services.validation.validation import ExamValidator, GradeValidator


class ExamService(ServiceBase):

    def __init__(self, database: AccessDatabase, logger: AppLogger,
                 journal: JournalService, parametres: ParametreService):
        super().__init__(database, logger, journal, parametres)
        self._examens = ExamRepository(database)
        self._bulletins = BulletinRepository(database)
        self._inscriptions = EnrollmentRepository(database)

    # ----- Sessions -----

    def list_sessions(self, id_annee: Optional[int] = None) -> List[SessionExam]:
        try:
            return self._examens.list_sessions(id_annee)
        except Exception as ex:
            self.logger.error("Liste des sessions d’examen", ex)
            return []

    def get_session(self, id_session: int) -> Optional[SessionExam]:
        try:
            return self._examens.get_session(id_session)
        except Exception as ex:
            self.logger.error("Lecture d’une session", ex)
            return None

    def create_session(self, session: SessionExam, code_utr: str) -> ResultValue[int]:
        if session.id_annee is None:
            from ltadmin.repositories.admin_repository import AdminRepository
            annee = AdminRepository(self.db).get_annee_active()
            if annee is None:
                return ResultValue.fail(
                    "Aucune année scolaire active : activez-en une dans "
                    "l’administration.", "GESTION")
            session.id_annee = annee.id_annee
        validation = ExamValidator.validate_session(session)
        if not validation.is_valid:
            return self.invalid_value(validation)
        try:
            if session.cloturee is None:
                session.cloturee = False
            session.id_session = self._examens.insert_session(session)
            self.journal.log_creation(
                code_utr, Tables.SESSION_EXAM, session.id_session, session.libelle)
            return ResultValue.ok(session.id_session, "Session d’examen créée.")
        except Exception as ex:
            return self.failure_value("Création d’une session", ex)

    def update_session(self, session: SessionExam, code_utr: str) -> Result:
        if session.id_session is None:
            return Result.fail("Session introuvable.", "INTROUVABLE")
        validation = ExamValidator.validate_session(session)
        if not validation.is_valid:
            return self.invalid(validation)
        try:
            updated = self._examens.update_session(session)
            if updated == 0:
                return Result.fail("Session introuvable.", "INTROUVABLE")
            self.journal.log_modification(
                code_utr, Tables.SESSION_EXAM, session.id_session, session.libelle)
            return Result.ok("Session enregistrée.")
        except Exception as ex:
            return self.failure("Modification d’une session", ex)

    def set_session_cloturee(self, id_session: int, cloturee: bool, code_utr: str) -> Result:
        try:
            session = self._examens.get_session(id_session)
            if session is None:
                return Result.fail("Session introuvable.", "INTROUVABLE")
            self._examens.set_session_cloturee(id_session, cloturee)
            self.journal.log_operation(
                code_utr,
                "CLOTURE_SESSION" if cloturee else "REOUVERTURE_SESSION",
                Tables.SESSION_EXAM, id_session, session.libelle)
            return Result.ok(
                "Session clôturée : les notes d’examen sont figées." if cloturee
                else "Session rouverte.")
        except Exception as ex:
            return self.failure("Clôture d’une session", ex)

    def delete_session(self, id_session: int, code_utr: str) -> Result:
        try:
            session = self._examens.get_session(id_session)
            if session is None:
                return Result.fail("Session introuvable.", "INTROUVABLE")
            self._examens.delete_session(id_session)
            self.journal.log_suppression(
                code_utr, Tables.SESSION_EXAM, id_session, session.libelle)
            return Result.ok("Session supprimée.")
        except Exception as ex:
            if error_helper.is_foreign_key_violation(ex):
                return Result.fail(
                    "Suppression impossible : des épreuves et des notes d’examen "
                    "sont rattachées à cette session.", "LIAISON")
            return self.failure("Suppression d’une session", ex)

    # ----- Épreuves -----

    def list_epreuves(self, id_session: Optional[int] = None,
                      id_classe: Optional[int] = None) -> List[EpreuveDetail]:
        try:
            return self._examens.list_epreuves(id_session, id_classe)
        except Exception as ex:
            self.logger.error("Liste des épreuves", ex)
            return []

    def get_epreuve(self, id_epreuve: int) -> Optional[Epreuve]:
        try:
            return self._examens.get_epreuve(id_epreuve)
        except Exception as ex:
            self.logger.error("Lecture d’une épreuve", ex)
            return None

    def get_epreuve_detail(self, id_epreuve: int) -> Optional[EpreuveDetail]:
        try:
            return self._examens.get_epreuve_detail(id_epreuve)
        except Exception as ex:
            self.logger.error("Lecture d’une épreuve", ex)
            return None

    def create_epreuve(self, epreuve: Epreuve, code_utr: str) -> ResultValue[int]:
        if epreuve.bareme is None or epreuve.bareme <= 0:
            epreuve.bareme = self.parametres.bareme_defaut
        if epreuve.coefficient is None or epreuve.coefficient <= 0:
            epreuve.coefficient = 1.0
        validation = ExamValidator.validate_epreuve(epreuve)
        if not validation.is_valid:
            return self.invalid_value(validation)
        try:
            session = self._examens.get_session(epreuve.id_session)
            if session is None:
                return ResultValue.fail("Session d’examen introuvable.", "INTROUVABLE")
            epreuve.id_epreuve = self._examens.insert_epreuve(epreuve)
            self.journal.log_creation(
                code_utr, Tables.EPREUVE, epreuve.id_epreuve,
                f"Épreuve {epreuve.code_matiere}")
            return ResultValue.ok(epreuve.id_epreuve, "Épreuve créée.")
        except Exception as ex:
            return self.failure_value("Création d’une épreuve", ex)

    def update_epreuve(self, epreuve: Epreuve, code_utr: str) -> Result:
        if epreuve.id_epreuve is None:
            return Result.fail("Épreuve introuvable.", "INTROUVABLE")
        validation = ExamValidator.validate_epreuve(epreuve)
        if not validation.is_valid:
            return self.invalid(validation)
        try:
            updated = self._examens.update_epreuve(epreuve)
            if updated == 0:
                return Result.fail("Épreuve introuvable.", "INTROUVABLE")
            self.journal.log_modification(
                code_utr, Tables.EPREUVE, epreuve.id_epreuve, epreuve.code_matiere)
            return Result.ok("Épreuve enregistrée.")
        except Exception as ex:
            return self.failure("Modification d’une épreuve", ex)

    def delete_epreuve(self, id_epreuve: int, code_utr: str) -> Result:
        try:
            epreuve = self._examens.get_epreuve(id_epreuve)
            if epreuve is None:
                return Result.fail("Épreuve introuvable.", "INTROUVABLE")
            self._examens.delete_epreuve(id_epreuve)
            self.journal.log_suppression(
                code_utr, Tables.EPREUVE, id_epreuve, epreuve.code_matiere)
            return Result.ok("Épreuve supprimée.")
        except Exception as ex:
            if error_helper.is_foreign_key_violation(ex):
                return Result.fail(
                    "Suppression impossible : des notes d’examen sont rattachées "
                    "à cette épreuve.", "LIAISON")
            return self.failure("Suppression d’une épreuve", ex)

    # ----- Saisie des notes d'examen -----

    def get_grille_saisie(self, id_epreuve: int) -> List[NoteSaisieRow]:
        try:
            epreuve = self._examens.get_epreuve(id_epreuve)
            if epreuve is None or epreuve.id_classe is None:
                return []
            return self._examens.list_grille_saisie(id_epreuve, epreuve.id_classe)
        except Exception as ex:
            self.logger.error("Grille de saisie des notes d’examen", ex)
            return []

    def upsert_note(self, id_epreuve: int, id_inscription: int,
                    valeur: Optional[float], absent: bool,
                    copie_num: Optional[str], code_utr: str) -> Result:
        """Enregistre la note d'examen d'un inscrit (0 ≤ note ≤ barème)."""
        try:
            epreuve = self._examens.get_epreuve(id_epreuve)
            if epreuve is None:
                return Result.fail("Épreuve introuvable.", "INTROUVABLE")
            session = self._examens.get_session(epreuve.id_session)
            if session is not None and session.cloturee is True:
                return Result.fail(
                    "Session clôturée : les notes d’examen sont figées.", "GESTION")
            bareme = epreuve.bareme if epreuve.bareme is not None \
                else self.parametres.bareme_defaut
            validation = GradeValidator.validate_note(valeur, absent, bareme)
            if not validation.is_valid:
                return self.invalid(validation)
            if not copie_num or not copie_num.strip():
                copie_num = None
            elif len(copie_num) > 40:
                return Result.fail("N° de copie : 40 caractères maximum.", "VALIDATION")

            existing = self._examens.get_note_examen(id_epreuve, id_inscription)
            if existing is None:
                id_note = self._examens.insert_note_examen(NoteExamen(
                    id_epreuve=id_epreuve,
                    id_inscription=id_inscription,
                    valeur_note=None if absent else valeur,
                    absent=absent,
                    copie_num=copie_num,
                    date_saisie=_dt.datetime.now(),
                    code_utr=code_utr))
                self.journal.log_creation(
                    code_utr, Tables.NOTE_EXAMEN, id_note,
                    f"Épreuve {id_epreuve} / inscrit {id_inscription}")
            else:
                existing.valeur_note = None if absent else valeur
                existing.absent = absent
                existing.copie_num = copie_num
                existing.date_saisie = _dt.datetime.now()
                existing.code_utr = code_utr
                self._examens.update_note_examen(existing)
                self.journal.log_modification(
                    code_utr, Tables.NOTE_EXAMEN, existing.id_note_ex,
                    f"Épreuve {id_epreuve} / inscrit {id_inscription}")
            return Result.ok("Note d’examen enregistrée.")
        except Exception as ex:
            return self.failure("Saisie d’une note d’examen", ex)

    # ----- Résultats finaux -----

    def list_resultats(self, id_classe: int) -> List[ResultatFinal]:
        try:
            return self._bulletins.list_by_classe(id_classe)
        except Exception as ex:
            self.logger.error("Liste des résultats finaux", ex)
            return []

    def get_resultat(self, id_inscription: int) -> Optional[ResultatFinal]:
        try:
            return self._bulletins.get_by_inscription(id_inscription)
        except Exception as ex:
            self.logger.error("Lecture d’un résultat final", ex)
            return None

    def generer_resultats(self, id_classe: int, id_session: int,
                          code_utr: str) -> ResultValue[int]:
        """Calcule les résultats finaux d'une classe (idempotent).

        MOY_CC = moyenne des bulletins ; MOY_EXAM = Σ(note/bareme×20×coef)/Σcoef ;
        MOYENNE_GEN = CC×POIDS_CC/100 + EXAM×POIDS_EXAMEN/100 (une partie absente
        → l'autre seule ; aucune → NULL). Décision ADMIS/AJOURNÉ selon MOY_ADMISSION.
        """
        try:
            inscrits = self._inscriptions.list_by_classe(id_classe)
            if not inscrits:
                return ResultValue.fail("Aucun étudiant inscrit dans cette classe.", "GESTION")

            poids_cc = self.parametres.poids_cc
            poids_examen = self.parametres.poids_examen
            moy_admission = self.parametres.moy_admission
            grille = self._bulletins.list_mentions()

            # Notes d'examen ramenées sur 20 et pondérées par coefficient.
            examens: dict = {}
            for row in self._examens.list_notes_ponderes(id_classe, id_session):
                id_inscription = row.get("ID_INSCRIPTION")
                if id_inscription is None:
                    continue
                accum = examens.setdefault(id_inscription, {"points": 0.0, "coef": 0.0})
                absent = bool(row.get("ABSENT"))
                note = row.get("VALEUR_NOTE")
                if absent or note is None:
                    continue
                bareme = row.get("BAREME") or 20.0
                coefficient = row.get("COEFFICIENT") or 1.0
                if bareme <= 0 or coefficient <= 0:
                    continue
                accum["points"] += (float(note) / float(bareme)) * 20.0 * float(coefficient)
                accum["coef"] += float(coefficient)

            generes = 0
            valeurs: list = []
            with self.db.begin_transaction():
                for inscrit in inscrits:
                    # Idempotent : une seule ligne de résultat par inscription.
                    self._bulletins.delete_by_inscription(inscrit.id_inscription)

                    moyennes_bulletins = self._bulletins.list_moyennes_by_inscription(
                        inscrit.id_inscription)
                    moy_cc = (sum(moyennes_bulletins) / len(moyennes_bulletins)
                              if moyennes_bulletins else None)

                    accum = examens.get(inscrit.id_inscription)
                    moy_exam = (accum["points"] / accum["coef"]
                                if accum and accum["coef"] > 0 else None)

                    if moy_cc is not None and moy_exam is not None:
                        moyenne_gen = (moy_cc * poids_cc / 100.0
                                       + moy_exam * poids_examen / 100.0)
                    elif moy_cc is not None:
                        moyenne_gen = moy_cc
                    elif moy_exam is not None:
                        moyenne_gen = moy_exam
                    else:
                        moyenne_gen = None

                    admis = moyenne_gen is not None and moyenne_gen >= moy_admission
                    mention = trouver(grille, moyenne_gen)

                    resultat = ResultatFinal(
                        id_inscription=inscrit.id_inscription,
                        moy_cc=moy_cc,
                        moy_exam=moy_exam,
                        moyenne_gen=moyenne_gen,
                        mention=mention.mention if mention else None,
                        decision="ADMIS" if admis else "AJOURNE",
                        credit_valide=None,
                        date_delib=_dt.datetime.now(),
                        observation=None,
                    )
                    self._bulletins.insert_resultat(resultat)
                    generes += 1
                    if moyenne_gen is not None:
                        valeurs.append((inscrit.id_inscription, moyenne_gen))

            rangs = rangs_competition(valeurs)
            for id_inscription, _ in valeurs:
                resultat = self._bulletins.get_by_inscription(id_inscription)
                if resultat is not None:
                    resultat.rang = rangs.get(id_inscription, 0)
                    self._bulletins.update_resultat_rang(resultat)

            self.journal.log_operation(
                code_utr, "GENERATION_RESULTATS", Tables.RESULTAT_FINAL, id_classe,
                f"{generes} résultat(s) calculé(s).")
            return ResultValue.ok(generes, f"{generes} résultat(s) final(aux) calculé(s).")
        except Exception as ex:
            return self.failure_value("Calcul des résultats finaux", ex)
